### PR TITLE
[Merged by Bors] - Support tls encryption and authentication

### PR DESCRIPTION
**Pull request by maltesander**
# Description

- Added TLS encryption
- Added TLS Authenication

Currently CRD looks like:
```
apiVersion: druid.stackable.tech/v1alpha1
kind: DruidCluster
metadata:
  name: derby-druid
spec:
  version: 24.0.0-stackable0.1.0
  clusterConfig:
    authentication:
      tls:
        authenticationClass: druid-mtls-authentication-class
    deepStorage:
      hdfs:
        configMapName: druid-hdfs
        directory: /druid
    metadataStorageDatabase:
      dbType: derby
      connString: jdbc:derby://localhost:1527/var/druid/metadata.db;create=true
      host: localhost
      port: 1527
    tls:
      secretClass: tls
    zookeeperConfigMapName: druid-znode
```

Not quite happy for the authentication part, basically the same information is repeated in the referenced AuthenticationClass. 

Still mergeable for now, we need to come up with a better way first.

Test: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/druid-operator-it-custom/37/

fixes https://github.com/stackabletech/druid-operator/issues/6

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: docs/modules/ROOT/pages/usage.adoc [hunk 1]
```diff
@@ -242,26 +240,90 @@ Druid can use HDFS as a backend for deep storage:
...
+----
+<1> Name of the `SecretClass` that is used to encrypt communication.
+
+IMPORTANT: A Stackable Druid cluster is always encrypted per default. In order to disable this default behavior you can set `spec.clusterConfig.tls: null`.
+
+=== Authentication
+
+==== TLS
+
+The access to the Druid cluster can be limited by configuring client authentication (mutual TLS) for all participants. This means that clients (e.g. the Browser or a Broker) have to authenticate themselves with valid certificates in order to communicate with the Druid cluster.
```
Comment (razvan) [inline docs/modules/ROOT/pages/usage.adoc:276]: It's not clear if this also enables internal tls communication between Druid servers or not. From the Druid docs, I understand that TLS is an all or nothing option. Also the operator code uses `tls` and `authentication` interchangeably. More clarification is needed.

**Review by razvan — COMMENTED** [review]
There is lot going on in here but it look very good in general
1. The new example/tls is missing the Zookeeper resources
2. Some additional (unit and integration) tests would help.
3. The docs could be more clear what the difference between setting `tls` and `authentication` is.
I haven't run the getting started code.

File: rust/operator-binary/src/druid_controller.rs [hunk 2]
```diff
@@ -247,9 +253,34 @@ pub async fn reconcile_druid(druid: Arc<DruidCluster>, ctx: Arc<Ctx>) -> Result<
...
+
+    let mut tls_authentication_config = None;
+    let mut authentication_config = vec![];
+    // Remove a TLS provider if available. The TLS provider must be treated in combination
+    // with TLS encryption. Retain all other authentication providers.
+    for auth_conf in resolved_authentication_config.into_iter() {
+        if auth_conf.is_tls_auth() {
+            tls_authentication_config = Some(auth_conf);
+        } else {
+            authentication_config.push(auth_conf);
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/druid_controller.rs:269]: Yeah thats basically a pre step for LDAP or other authentication mechanisms (which will be in here then).

File: rust/crd/src/lib.rs [hunk 3]
```diff
@@ -283,8 +328,13 @@ impl DruidCluster {
                     String::from(EXT_OPA_AUTHORIZER),
                     String::from(EXT_HDFS),
                 ];
+
+                if self.tls_enabled() {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs:332]: Yes

File: examples/tls/tls-druid-cluster.yaml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        configMapName: druid-hdfs
+        directory: /druid
+    metadataStorageDatabase:
+      dbType: derby
+      connString: jdbc:derby://localhost:1527/var/druid/metadata.db;create=true
+      host: localhost
+      port: 1527
+    tls:
+      secretClass: tls
+    zookeeperConfigMapName: druid-znode
```
[the remark this answers is not in the document]
Comment (maltesander) [reply examples/tls/tls-druid-cluster.yaml:93]: Good catch! Done https://github.com/stackabletech/druid-operator/pull/333/commits/c148abdf96ce21948531d66c95b1bcfaa244f51a

File: rust/crd/src/authentication.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum DruidAuthenticationConfig {
+    Tls(TlsAuthenticationProvider),
+}
+
+impl DruidAuthenticationConfig {
+    pub fn is_tls_auth(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/authentication.rs:92]: No its a different question. `tls_enabled` signals that we need some sort of TLS. This affects ports, config settings, volume mounts etc. (can be encryption or authentication).
The `is_tls_auth` is used to split up the vector of authentication mechanism and bundle TLS encryption and authentication in an extra struct to better process this from there on.

File: rust/operator-binary/src/druid_controller.rs [hunk 6]
```diff
@@ -719,12 +698,83 @@ fn build_rolegroup_statefulset(
...
+
+                            let volume = VolumeBuilder::new(&volume_name)
+                                .ephemeral(
+                                    SecretOperatorVolumeSourceBuilder::new(secret_class).build(),
+                                )
+                                .build();
+                            pb.add_volume(volume);
+                            cb_druid.add_volume_mount(
+                                &volume_name,
+                                format!("{CERTS_DIR}{volume_name}"),
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/druid_controller.rs:764]: I dont get it? It is `CERTS_DIR`?

[the remark this answers is not in the document]
Comment (razvan) [reply rust/operator-binary/src/druid_controller.rs:764]: In crd/src/lib.rs:271:
```
shell_cmd.push(format!("keytool -importcert -file /stackable/certificates/{secret_class}-tls-certificate/ca.crt -alias stackable-{secret_class} -keystore {STACKABLE_TRUST_STORE} -storepass {STACKABLE_TRUST_STORE_PASSWORD} -              noprompt"));
```
should use this constant.

File: rust/crd/src/tls.rs [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+            config.insert(ENABLE_TLS_PORT.to_string(), Some("true".to_string()));
+            config.insert(
+                TLS_PORT.to_string(),
+                Some(role.get_https_port().to_string()),
+            );
+        }
+    }
+
+    /// Add required TLS ports, trust/key store properties
+    pub fn add_tls_config_properties(
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/tls.rs:201]: Done https://github.com/stackabletech/druid-operator/pull/333/commits/a1ff84441aa35307d57af459cc51fba59ff722c8

File: docs/modules/ROOT/pages/usage.adoc (hunk 1 above)
Comment (maltesander) [reply docs/modules/ROOT/pages/usage.adoc]: Tried to improve it here https://github.com/stackabletech/druid-operator/pull/333/commits/d0ab6ab77bcc8c1ad5dfe8383ea578731b8b342e

File: rust/crd/src/tls.rs [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+        &self,
+        config: &mut BTreeMap<String, Option<String>>,
+        role: &DruidRole,
+    ) {
+        if self.encryption.is_some() || self.authentication.is_some() {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/tls.rs:208]: Yeah it split of from the DruidCluster (in order to combine TLS encryption and TLS authentication which are separated in the CRD)

File: rust/operator-binary/src/druid_controller.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/druid_controller.rs]: Ah yeah done: https://github.com/stackabletech/druid-operator/pull/333/commits/08413cdfe24b4380929716b42087fca1a8c418de

File: rust/operator-binary/src/druid_controller.rs [hunk 9]
```diff
@@ -719,12 +698,83 @@ fn build_rolegroup_statefulset(
...
+    if let Some(s3_conn) = s3_conn {
+        if let Some(credentials) = &s3_conn.credentials {
+            pb.add_volume(credentials.to_volume("s3-credentials"));
+            cb_druid.add_volume_mount("s3-credentials", S3_SECRET_DIR_NAME);
+        }
+
+        if let Some(tls) = &s3_conn.tls {
+            match &tls.verification {
+                TlsVerification::None {} => return S3TlsNoVerificationNotSupportedSnafu.fail(),
+                TlsVerification::Server(server_verification) => {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/druid_controller.rs:750]: I switched the TLS tests to use S3 deep storage (with TLS) and added an ingestion check https://github.com/stackabletech/druid-operator/pull/333/commits/8e6bab3767c5a39cb69296522e645b6e37e32435

**Review by razvan — APPROVED** [review]
lgtm.

**Comment by maltesander** [comment]
Test: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/druid-operator-it-custom/38/

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**